src/control/server/ctl_storage_rpc.go

```diff
@@ -61,73 +70,118 @@ func newResponseState(inErr error, badStatus ctlpb.ResponseStatus, infoMsg strin
 
-	pbCtrlrs := make(proto.NvmeControllers, 0, len(inResp.Controllers))
-	if err := pbCtrlrs.FromNative(inResp.Controllers); err != nil {
-		return nil, err
+	for _, ei := range instances {
+		eReq := new(ctlpb.ScanNvmeReq)
+		*eReq = *req
+		ms, rs, err := cs.computeMetaRdbSize(ei, nsps)
+		if err != nil {
+			return nil, errors.Wrap(err, "computing meta and rdb size")
+		}
+		eReq.MetaSize, eReq.RdbSize = ms, rs
+
+		respEng, err := scanEngineBdevs(ctx, ei, eReq)
+		if err != nil {
+			err = errors.Wrapf(err, "instance %d", ei.Index())
+			if errLast == nil && len(instances) > 1 {
+				errLast = err // Save err to preserve partial results.
+				cs.log.Error(err.Error())
+				continue
+			}
+			return nil, err // No partial results to save so fail.
+		}
+		resp.Ctrlrs = append(resp.Ctrlrs, respEng.Ctrlrs...)
 	}
 
-	// trim unwanted fields so responses can be coalesced from hash map
-	for _, pbc := range pbCtrlrs {
+	// If one engine succeeds and one other fails, error is embedded in the response.
+	resp.State = newResponseState(errLast, ctlpb.ResponseStatus_CTL_ERR_NVME, "")
+
+	return resp, nil
+}
+
+// Trim unwanted fields so responses can be coalesced from hash map when returned from server.
+func bdevScanTrimResults(req *ctlpb.ScanNvmeReq, resp *ctlpb.ScanNvmeResp) *ctlpb.ScanNvmeResp {
+	if resp == nil {
+		return nil
+	}
+	for _, pbc := range resp.Ctrlrs {
 		if !req.GetHealth() {
 			pbc.HealthStats = nil
 		}
 		if !req.GetMeta() {
 			pbc.SmdDevices = nil
 		}
 		if req.GetBasic() {
-			stripNvmeDetails(pbc)
+			pbc.Serial = ""
+			pbc.Model = ""
+			pbc.FwRev = ""
 		}
 	}
 
-	outResp.Ctrlrs = pbCtrlrs
-
-	return outResp, nil
+	return resp
 }
 
-// scanBdevs updates transient details if health statistics or server metadata
-// is requested otherwise just retrieves cached static controller details.
-func (c *ControlService) scanBdevs(ctx context.Context, req *ctlpb.ScanNvmeReq, nsps []*ctlpb.ScmNamespace) (*ctlpb.ScanNvmeResp, error) {
+// Return NVMe device details. The scan method employed depends on whether the engines are running
+// or not. If running, scan over dRPC. If not running then use engine's storage provider.
+func bdevScan(ctx context.Context, cs *ControlService, req *ctlpb.ScanNvmeReq, nsps []*ctlpb.ScmNamespace) (resp *ctlpb.ScanNvmeResp, err error) {
 	if req == nil {
-		return nil, errors.New("nil bdev request")
+		return nil, errors.New("nil request")
 	}
 
-	var bdevsInCfg bool
-	for _, ei := range c.harness.Instances() {
-		if ei.GetStorage().HasBlockDevices() {
-			bdevsInCfg = true
+	cfgBdevs := getBdevCfgsFromSrvCfg(cs.srvCfg).Bdevs()
+	hasBdevs := cfgBdevs.Len() != 0
+
+	// Note the potential window where engines are started but not yet ready to respond. In this
+	// state there is a possibility that neither scan mechanism will work because devices have
+	// been claimed by SPDK but details are not yet available over dRPC. Here it is assumed that
+	// devices are claimed as soon the engines are started.
```

Comment:
What happens in this case? Does the drpc request return -DER_AGAIN?

Reply:
I see potentially another race, where the code here sees no engines started, but then the engines start before the global bdev scan executes. What does that case look like to the user?

Reply:
1. In the situation where SPDK has claimed the controllers for the engine process but dRPC is not yet ready for comms, an "engine not ready" error should be returned. The window should be <<1s.
2. In the situation where this check results in a global bdev scan but by the time the scan executes the controllers have been claimed by the engine process, I think no results will be returned.

In terms of how to handle the above 2 scenarios I suggest the following:
- If "engine not ready" we issue a similar fault suggesting a retry.
- If no bdevs are found but some are expected (as bdev_list has been populated in config), fault suggesting retry is given.

I would prefer not to make it more complicated than this at the current stage to avoid the risk of pre-empting a solution for something that isn't a problem in practice.
Does the above suggestion sound acceptable for now?

Reply:
Sounds good to me. I just wanted to make sure we weren't hiding an error or returning weird/confusing results.

Reply:
I think that Kris raises a valid concern, particularly in the context of CI testing -- we don't want to introduce a new source of intermittent failures. Why should the retry error be exposed to the user vs. just handling the retry in code?

Reply:
Changes to implement retries when scanning over provider and no results are returned and also to wait for drpc to be ready before scanning when engine is running have been applied in this PR.